src/nsrt_learning/nsrt_learning_main.py

```diff
@@ -224,3 +256,88 @@ def _learn_pnad_samplers(pnads: List[PartialNSRTAndDatastore],
+                       option_specs: List[OptionSpec]) -> bool:
+    """Function to check whether a given set of operators and predicates
+    preserves harmlessness over demonstrations on some number of training
+    tasks."""
+
+    for ll_traj, hl_traj in pruned_atom_data:
+        if not ll_traj.is_demo:
+            continue
+        traj_goal = train_tasks[ll_traj.train_task_idx].goal
+        plan_preserved = check_single_plan_preservation(
+            ll_traj, hl_traj, traj_goal, init_preds, strips_ops, option_specs)
+        if not plan_preserved:
+            return False
+
+    return True
+
+
+def check_single_plan_preservation(ll_traj: LowLevelTrajectory,
+                                   hl_traj: List[Set[GroundAtom]],
+                                   traj_goal: Set[GroundAtom],
+                                   init_preds: Set[Predicate],
+                                   strips_ops: List[STRIPSOperator],
+                                   option_specs: List[OptionSpec]) -> bool:
+    """Function to check whether a given set of operators and predicates
+    preserves a sigle training trajectory."""
+    init_atoms = utils.abstract(ll_traj.states[0], init_preds)
+    objects = set(ll_traj.states[0])
+    ground_nsrts, _ = task_plan_grounding(init_atoms, objects, strips_ops,
+                                          option_specs)
+    heuristic = utils.create_task_planning_heuristic(
+        CFG.sesame_task_planning_heuristic, init_atoms, traj_goal,
+        ground_nsrts, init_preds, objects)
+
+    def _check_goal(state: Tuple[FrozenSet[GroundAtom], int]) -> bool:
+        return traj_goal.issubset(state[0])
+
+    def _get_successor_with_correct_option(
+        searchnode_state: Tuple[FrozenSet[GroundAtom], int]
+    ) -> Iterator[Tuple[_GroundNSRT, Tuple[FrozenSet[GroundAtom], int],
+                        float]]:
+        state = searchnode_state[0]
+        idx_into_traj = searchnode_state[1]
+
+        if idx_into_traj <= len(ll_traj.actions) - 1:
+            assert ll_traj.actions[idx_into_traj].has_option()
+            gt_option = ll_traj.actions[idx_into_traj].get_option()
+            expected_next_hl_state = hl_traj[idx_into_traj + 1]
+
+            for applicable_nsrt in utils.get_applicable_operators(
+                    ground_nsrts, state):
+                # NOTE: we check that the ParameterizedOptions are equal before
+                # attempting to ground because otherwise, we might
+                # get a parameter mismatch and trigger an AssertionError
+                # during grounding.
+
+                if applicable_nsrt.option == gt_option.parent:
+                    if applicable_nsrt.option_objs == gt_option.objects:
+                        next_hl_state = utils.apply_operator(
+                            applicable_nsrt, set(state))
+                        # Here, we check whether all atoms that differ
+                        # between next_hl_state and state are part of
```

Comment:
are you actually checking that they're all part of the operator's side predicates? it looks like you're just checking issubset, which is a weaker condition right?

Reply:
indeed - this comment is outdated!